generators/chipyard/src/main/scala/IOBinders.scala

```diff
@@ -364,40 +408,53 @@ class WithTieOffL2FBusAXI extends OverrideIOBinder({
   }
 })
 
-// TODO we need to rethink what "Tie-off-debug" means. The current system punches out
-// excessive IOs.
-class WithTiedOffDebug extends OverrideIOBinder({
+class WithSimDebug extends OverrideIOBinder({
   (system: HasPeripheryDebugModuleImp) => {
-    val (psdPort, resetctrlOpt, debugPortOpt, ioCells) =
-      AddIOCells.debug(system.psd, system.resetctrl, system.debug)(system.p)
+    val (ports, iocells) = AddIOCells.debug(system)(system.p)
     val harnessFn = (th: HasHarnessSignalReferences) => {
-      Debug.tieoffDebug(debugPortOpt, resetctrlOpt, Some(psdPort))(system.p)
-      // tieoffDebug doesn't actually tie everything off :/
-      debugPortOpt.foreach { d =>
-        d.clockeddmi.foreach({ cdmi => cdmi.dmi.req.bits := DontCare; cdmi.dmiClock := th.harnessClock })
-        d.dmactiveAck := DontCare
-        d.clock := th.harnessClock // TODO fix: This should be driven from within the chip
+      val dtm_success = Wire(Bool())
+      when (dtm_success) { th.success := true.B }
+      ports.map {
+        case d: ClockedDMIIO =>
+          val dtm = Module(new SimDTM()(system.p)).connect(th.harnessClock, th.harnessReset.asBool, d, dtm_success)
+        case j: JTAGIO =>
+          val jtag = Module(new SimJTAG(tickDelay=3)).connect(j, th.harnessClock, th.harnessReset.asBool, ~(th.harnessReset.asBool), dtm_success)
+        case _ =>
+          require(false, "We only support DMI or JTAG simulated debug connections")
       }
       Nil
     }
-    Seq((Seq(psdPort) ++ resetctrlOpt ++ debugPortOpt.toSeq, Nil, Some(harnessFn)))
+    Seq((ports, iocells, Some(harnessFn)))
   }
 })
 
-// TODO we need to rethink what this does. The current system punches out excessive IOs.
-// Some of the debug clock/reset should be driven from on-chip
-class WithSimDebug extends OverrideIOBinder({
+class WithTiedOffDebug extends OverrideIOBinder({
```

Comment:
Are we going to add an option that actually removes the debug IOs from going off-chip? This currently just makes the debug unit unusable without making it so that CAD tools would be able to prune portions of it.

Reply:
What do you mean? 
A system with no DebugIOs going off chip can be built by setting `DebugModuleKey` to `None`. Do we want to support a case where there is a hardware debug module, but no IOs to interact with it?

Reply:
I don't think we should expect CAD tool pruning to be the proper method for eliminating unwanted hardware. In general, IOs which appear in ChipTop, should be expected to be IOs desired on the chip.

Reply:
Ok I didn't realize you could set `DebugModuleKey` to `None`.
The use case for this function then is to support implementations that have a debug module but don't want to use it now? What does this provide? Can't we just always connect it and let the outer level simulation decide which to drive at runtime?

Reply:
> Ok I didn't realize you could set DebugModuleKey to None.
This is untested, but I'll verify this works. I assume we don't need to prioritize this use case? 

> Can't we just always connect it and let the outer level simulation decide which to drive at runtime?
If we attach both SimDTM and SimSerial in the TestHarness, than both bringup methods will read the fesvr args and attempt to communicate with the DUT, which causes undefined behavior. The primary use case is to support systems with both clockedDMI and serial interfaces.

Perhaps we should default to having the debug module use JTAG, instead of clockedDMI. In that case, `WithTiedOffDebug` would not be useful, since we can support simulations with both SimSerial and SimJTAG
.

Reply:
Hmm. I wonder if we can make SimDTM or SimSerial able to cooperate with the other, and then eliminate this tieOff altogether.

Reply:
That may be challenging, as they instantiate separate `htif_t` objects. We could possibly make this work, but should we? We don't build real chips with the clockedDMI interface, so perhaps we should just make the default case more accurate, with JTAG+Serial, instead of clockedDMI+Serial.

Reply:
Yeah I'm happy making that the default. I think it might be nice to `assert` that we don't support DMI+Serial.

Reply:
Its a bit tricky to `assert` this actually, since the IOBinders are by design isolated from each other.
I think its ok to not assert on this for now, since the dmiRocketConfig provides an example on supporting a DMI+Serial system, and since most people seem to want either serial interface or JTAG.

Reply:
Ok. How does that config/system figure out which interface is used for hosted-bringup/testing?

Reply:
It has `WithTiedOffSerial`, which does not attach the `SimSerial` widget.

Reply:
Ok. I guess whoever really uses that config first can figure out the software issues.

Reply:
To be clear, the included dmiRocketConfig works. What doesn't work is a config with
```
new WithSimSerial ++
new WithSimDebug ++
new WithDMIDTM
```

Reply:
yeah I understand. The software I'm talking about is the FESVR side of things that would let you use serial as off-chip memory while using the DMI to do interactive debugging.